File: app/src/main/java/ke/choxxy/wordsearch/di/modules/AppModule.java
package ke.choxxy.wordsearch.di.modules;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import ke.choxxy.wordsearch.features.ViewModelFactory;
import ke.choxxy.wordsearch.data.GameDataSource;
import ke.choxxy.wordsearch.data.GameThemeRepository;
import ke.choxxy.wordsearch.data.WordDataSource;
import ke.choxxy.wordsearch.features.gamehistory.GameHistoryViewModel;
import ke.choxxy.wordsearch.features.gameover.GameOverViewModel;
import ke.choxxy.wordsearch.features.gameplay.GamePlayViewModel;
import ke.choxxy.wordsearch.features.mainmenu.MainMenuViewModel;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by abdularis on 18/07/17.
 */

@Module
public class AppModule {

    private Application mApp;

    public AppModule(Application application) {
        mApp = application;
    }

    @Provides
    @Singleton
    Context provideContext() {
        return mApp;
    }

    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    ViewModelFactory provideViewModelFactory(GameDataSource gameDataSource,
                                             WordDataSource wordDataSource) {
        return new ViewModelFactory(
                new GameOverViewModel(gameDataSource),
                new GamePlayViewModel(gameDataSource, wordDataSource),
                new MainMenuViewModel(new GameThemeRepository()),
                new GameHistoryViewModel(gameDataSource)
        );
    }
}

File: app/src/main/java/ke/choxxy/wordsearch/di/AppModule.kt
package ke.choxxy.wordsearch.di

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import ke.choxxy.wordsearch.ViewModelFactory
import ke.choxxy.wordsearch.data.GameThemeRepository
import ke.choxxy.wordsearch.data.WordDataSource
import ke.choxxy.wordsearch.data.sqlite.DbHelper
import ke.choxxy.wordsearch.data.sqlite.GameDataSQLiteDataSource
import ke.choxxy.wordsearch.gamehistory.GameHistoryViewModel
import ke.choxxy.wordsearch.gameover.GameOverViewModel
import ke.choxxy.wordsearch.gameplay.GamePlayViewModel
import ke.choxxy.wordsearch.mainmenu.MainMenuViewModel
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Provides
    @Singleton
    fun provideSharedPreferences(@ApplicationContext context: Context): SharedPreferences {
        return context.getSharedPreferences("word-search", MODE_PRIVATE)
    }

    @Provides
    @Singleton
    fun provideViewModelFactory(
        gameDataSource: GameDataSQLiteDataSource,
        wordDataSource: WordDataSource
    ): ViewModelFactory {
        return ViewModelFactory(
            GameOverViewModel(gameDataSource),
            GamePlayViewModel(gameDataSource, wordDataSource),
            MainMenuViewModel(GameThemeRepository()),
            GameHistoryViewModel(gameDataSource)
        )
    }

    @Provides
    @Singleton
    fun provideDbHelper(@ApplicationContext context: Context?): DbHelper {
        return DbHelper(context)
    }
}
